scRNA-seq-advanced/03-dataset_integration.Rmd

```diff
@@ -201,169 +293,238 @@ genes_per_sce <- sce_list %>%
+                       colour_by = "sample",
+                       # Include some point styling to help us see the points
+                       point_size = 0.5,
+                       point_alpha = 0.2) +
+  # Modify the legend key so its points are larger and easier to see
+  guides(colour = guide_legend(override.aes = list(size = 3, alpha = 1))) +
+  # Add a plot title
+  ggtitle("UMAP calculated on each library separately")
+```
+
+
+As we see in this UMAP, all libraries are centered at zero and all overlapping!
+This visual artifact can give the _incorrect impression_ that data is integrated - to be clear, this data is NOT integrated!
+
+For input to integration, we'll want the PCA scaling to consider normalized gene expression values from all libraries simultaneously, while also taking into consideration batch differences, so we'll need to recalculate PCA (and UMAP for visualization). 
+We'll also save these new reduced dimensions with different names, `merged_PCA` and `merged_UMAP`, to distinguish them from already-present `PCA` and `UMAP`.
+
+
+First, as usual, we'll determine the high-variance genes to use for PCA from the `merged_sce` object.
+For this, we'll also provide the argument `block = merged_sce$sample` when modeling gene variance, which tells `scran::modelGeneVar()` to first model variance separately for each batch and then combine those modeling statistics.
+
+```{r calc merged hv_genes}
+num_genes <- 2000
 
-```{r}
 # calculate high-variance genes to use for dimension reduction
 gene_variance <- scran::modelGeneVar(merged_sce,
                                      # specify the grouping/batch column:
                                      block = merged_sce$sample)
+
+# find the high-variance genes
 hv_genes <- scran::getTopHVGs(gene_variance,
-                              n = 2000)
+                              n = num_genes)
 ```
 
 
-Next, we need to calculate PCs. 
-Since we're calculating PCs specifically on a merged object, we will want to use a slightly different approach that respects the batch differences within. 
+To calculate the PCs themselves, we'll use an approach from the `batchelor` package, which is the R package that contains the `fastMNN` method.
+In particular, we'll use the `batchelor::multiBatchPCA()` function which accommodates the fact that we have separate batches in our SCE to perform the appropriate scaling we need.
+According to the documentation for this function, 
+> This function is roughly equivalent to `cbinding` all matrices in and performing PCA on the merged matrix. The main difference is that each sample is forced to contribute equally to the identification of the rotation vectors. Specifically, the mean vector used for centering is defined as the grand mean of the mean vectors within each batch. Each batch's contribution to the gene-gene covariance matrix is also divided by the number of cells in that batch. 
+
```

Comment:
I think having the documentation here is helpful, but it also might be nice to have a one sentence summary of what that means. I can see people having questions because they won't understand this description.

Reply:
Agreed: I think a link to the documentation is fine. I might just say that it prevents implicitly weighting by the size of the batch. Or that it corrects for the varying number of cells in each sample/batch.